lib/fs.js

```diff
@@ -2014,30 +2014,49 @@ function ReadStream(path, options) {
   this.closed = false;
 
   if (this.start !== undefined) {
-    if (typeof this.start !== 'number' || Number.isNaN(this.start)) {
+    if (typeof this.start !== 'number') {
       throw new ERR_INVALID_ARG_TYPE('start', 'number', this.start);
     }
-    if (this.end === undefined) {
-      this.end = Infinity;
-    } else if (typeof this.end !== 'number' || Number.isNaN(this.end)) {
-      throw new ERR_INVALID_ARG_TYPE('end', 'number', this.end);
+
+    if (Number.isNaN(this.start)) {
+      throw new ERR_OUT_OF_RANGE('start', 'not NaN', this.start);
     }
 
-    if (this.start > this.end) {
-      const errVal = `{start: ${this.start}, end: ${this.end}}`;
-      throw new ERR_OUT_OF_RANGE('start', '<= "end"', errVal);
+    if (this.start < 0) {
+      throw new ERR_OUT_OF_RANGE('start', 'positive', this.start);
+    }
+
+    if (!Number.isSafeInteger(this.start)) {
+      throw new ERR_OUT_OF_RANGE('start', 'safe integer', this.start);
     }
 
     this.pos = this.start;
   }
 
-  // Backwards compatibility: Make sure `end` is a number regardless of `start`.
-  // TODO(addaleax): Make the above typecheck not depend on `start` instead.
-  // (That is a semver-major change).
-  if (typeof this.end !== 'number')
+  if (this.end === undefined) {
     this.end = Infinity;
-  else if (Number.isNaN(this.end))
-    throw new ERR_INVALID_ARG_TYPE('end', 'number', this.end);
+  } else if (this.end !== Infinity) {
+    if (typeof this.end !== 'number') {
+      throw new ERR_INVALID_ARG_TYPE('end', 'number', this.end);
+    }
+
+    if (Number.isNaN(this.end)) {
+      throw new ERR_OUT_OF_RANGE('end', 'not NaN', this.end);
+    }
+
+    if (this.end < 0) {
+      throw new ERR_OUT_OF_RANGE('end', 'positive', this.end);
+    }
+
+    if (!Number.isSafeInteger(this.end)) {
+      throw new ERR_OUT_OF_RANGE('end', 'safe integer', this.end);
+    }
+
+    if (this.start !== undefined && this.start > this.end) {
+      const errVal = `{start: ${this.start}, end: ${this.end}}`;
+      throw new ERR_OUT_OF_RANGE('start', '<= "end"', errVal);
```

Comment:
Looking at this again I would also change this to:

```js
throw new ERR_OUT_OF_RANGE('start', `<= ${this.end}`, this.start);
```

This ends up as:

`The value of "start" is out of range. It must be <= ${this.end}. Received ${this.start}`

instead of:

`The value of "start" is out of range. It must be <= "end". Received {start: ${this.start}, end: ${this.end}}`

It just seems better to read for me. It is also the way how it is done in e.g. the `buffer` module.

But that is just a suggestion.

Reply:
Hmm, I'm not as inclined regarding this one.

Let's assume that the passed value of `end` is 4, and `start` is 6. The error message would be: `The value of "start" is out of range. It must be <= ${4}. Received ${2}`.

I bet there would be a section of developers, already coping with a huge language barrier, who would might to the conclusion that there's some magic value (4 in this case) which "start" needs to be smaller than.

What about: `The value of "start" is out of range. It must be <= "end" (here: 4). Received 6` instead?

Reply:
:+1: that works as well.